### ISSUE TITLE
ENH: Support low-latency sound card output

### ISSUE THREAD

**Issue by larsoner** [LABSN/expyfun#343]
@rkmaddox not sure if you have looked into libraries. Here are some potential options I found:

Best possibilities (upon brief inspection):

- Python: [pyo](https://github.com/belangeo/pyo) is what Psychopy says provides the best performance ([docs](http://ajaxsoundstudio.com/pyodoc/index.html))
- Python: [simpleaudio](https://github.com/hamiltron/py-simple-audio/)
- Python: [python-sounddevice](https://github.com/spatialaudio/python-sounddevice/)

Harder / less likely:
- C wrapper: [RtAudio has Python bindings](https://github.com/thestk/rtaudio/tree/master/contrib/python/pyrtaudio) updated 7 months ago -- need to check to make sure buffer operations are non-blocking. (RtAudio is what we used back in grad school!)
- C wrapper: [RtAudio also has other Python bindings](https://github.com/theintencity/py-audio), but last updated 1 year ago; does not work on Python 3
- C: [libsoundio](http://libsound.io/) doesn't have bindings, but if its interface is simple enough then a `ctypes` wrapper might be plausible

Unlikely:
- Python: [pyglet](http://pyglet.readthedocs.io/en/pyglet-1.3-maintenance/programming_guide/media.html) supports audio, but IIRC internally uses Python callbacks so wouldn't be safe enough (FWIW this is what we use by default for testing purposes currently)
- Python: [pyaudio](http://people.csail.mit.edu/hubert/pyaudio/docs/) appears to also use Python callbacks

**Comment by drammock** [comment]
I tried to use pyo instead of pyglet back when we created expyfun, but couldn't get it to work at the time. Not sure if it was me or pyo that failed (or both) but would be great if it has progressed, it looked so much more promising than pyglet.

**Comment by rkmaddox** [comment]
I have only looked a little into pyglet and it does not seem like it will do the job. In addition to callbacks, it also only allows one soundcard at a time, which is not ideal.

First impressions:
pyo looks great if it works.
simpleaudio requires py3k.
python-sounddevice looks like the simplest solution, but less powerful/flexible than pyo.

Didn't you also use RtAudio for your work with Cara? I remember it being a solid little library. But would prefer a python-specific solution since it is likeliest to work well in my mind (and forum posts about issues will be for python code, rather than C++).

**Comment by rkmaddox** [comment]
Also, pyo has this which is pretty cool: http://radiopyo.acaia.ca/

**Comment by drammock** [comment]
For an experiment run in the UW Phonetics lab, I've recently used `sounddevice` for *recording* audio, but not for playing out (I used expyfun/pyglet, because we didn't care about RT).  In case it's useful, the setup code is [here](https://github.com/drammock/gend-intel/blob/master/run-experiment.py#L47-L61) and the recording code is [here](https://github.com/drammock/gend-intel/blob/master/run-experiment.py#L173-L184) (the `try` block is to allow for arbitrary-length recording; it catches `KeyboardInterrupt`s to tell it when to close the input buffer).

**Comment by rkmaddox** [comment]
If we have access to the hardware audio buffers themselves, we could even out the jitter to be near zero, couldn't we? After all, it's the latency variation, not the lag itself, that is our problem. This white paper explains it well: https://www.rme-audio.de/old/english/techinfo/lola_latec.htm.

But if we know what time it is and what time the next buffer will turn over, then we should be able to zero-prepend our sound so that the latency is always some constant larger than 2 buffers. If I have it right in my head, we could guarantee a nearly jitterless latency with a length of three (or maybe it's two?) buffers.

**Comment by rkmaddox** [comment]
See the very simple example for sound-device [here](http://python-sounddevice.readthedocs.io/en/0.3.10/#callback-streams). It looks like it should actually be pretty easy to fix the latency. I will give it a quick shot and post a gist if possible. This library was stupidly simple to install.

**Comment by drammock** [comment]
not to send this off in yet another direction, but for possible future reference I just discovered a serious-looking real-time linux co-kernel called [xenomai](https://xenomai.org/)

**Comment by rkmaddox** [comment]
Pretty cool. Unfortunately, official sound card drivers are pretty scant for fancy soundcards (notably RME, which I am for sure going with). If we can get the buffer stuff working reliably, then it ought to be all right. Will keep thread posted.

**Comment by larsoner** [comment]
> But if we know what time it is and what time the next buffer will turn over, then we should be able to zero-prepend our sound so that the latency is always some constant larger than 2 buffers. If I have it right in my head, we could guarantee a nearly jitterless latency with a length of three (or maybe it's two?) buffers.

Yes but I always assumed it would be a PITA. If it's not, then we in principle don't even need the sound card to be particularly good to get zero(ish) latency because we can use standard size sound buffers instead of tiny ones. This would also reduce the risk of missing a buffer fill, which would be another added benefit.

Also, getting the buffer fill operations to occur *outside* of the Python kernel might not be so bad. A small Cython module that uses `nogil` to actually do the (effective) `memcpy` operation could even do it.

**Comment by larsoner** [comment]
While we're at it, we should just convert button presses to digital sound input bits, use a full duplex sound card, and get our input presses that way. They would be aligned to the sound output at the accuracy of the sound card's sample rate :)

**Comment by rkmaddox** [comment]
I think the button press recording might not be that difficult. Not sure though. It would require dedicated hardware, though.

**Comment by rkmaddox** [comment]
As for zero-prepending--it also might not be that difficult. Still could be though

**Comment by rkmaddox** [comment]
So there is a callback function (which you define in python) that is called in a separate thread by a running audio stream object. So we at least don't have to worry about that aspect of it.

What happens in that thread would probably be faster in C than python, but if all it's doing is playing / storing audio it shouldn't be too taxing for the CPU.

**Comment by larsoner** [comment]
> So there is a callback function (which you define in python) that is called in a separate thread by a running audio stream object. So we at least don't have to worry about that aspect of it.

That's good. With the GIL, the critical points then IIUC are:

1. When will the Python interpreter get around to executing the callback, and
2. How long will that callback occupy the interpreter (preventing other ops).

(1) becomes a non-issue with sufficiently large buffer sizes, and (2) I think can be mitigated by almost immediately delegating the buffer fill operation to a `nogil` C call. Actually from what I [just read](http://scipy-cookbook.readthedocs.io/items/ParallelProgramming.html#Threads), buffer[:] = mysound[:] in NumPy might already effectively be `nogil`, so we might not even need Cython to limit the GIL lock time.

> As for zero-prepending--it also might not be that difficult. Still could be though

I wouldn't worry about the buffer-delay compensation for now. Let's try to get this sound + PCM trigger stuff working at all (with any lib) to verify first that we can get:

1. Zero jitter, measured with a scope. Hard part: configuring the sound properly and getting the hardware set up properly. Otherwise the sound card should guarantee us this behavior.
2. No pops and clicks during playback. This should depend on buffer size (and system load), measured by listening and finding the min safe buffer size.
3. No delays in concurrent operations (i.e., check `nogil` effectiveness), which can probably be measured by:
    ```
    >>> start_sound()
    ... t0 = time.time()
    ... while True:
    ...     times.append(time.time())
    <ctrl-c after sound finishes>
    >>> print(np.max(np.diff(times))
    ```
    If this value is much smaller than `buffer_size / fs`, then we know the callback can't be hogging the GIL, and that we can effectively asynchronously execute.

WDYT about this plan? @rkmaddox do you have time to try it? @drammock and I could at least maybe work on (2) and (3), but (1) would need to be done with your hardware setup I think. (At least until we're convinced it all works and we all switch!)

**Comment by rkmaddox** [comment]
That seems like a good plan. The only downside is I may end up buying way more sound card than is necessary, since without knowing if we can handle the jitter thing, then I want to be sure I can get as small a buffer as possible. I am still trying to figure out which system to go with.

Re: 1, the boards should be ordered next week. However, Lenny has already tested this with their setup and the latency jitter is always < 1 sample.

Re: 2, yep will depend on lots of things. Sound card, host computer, frame size, block size, etc. But there will clearly be some settings that yield that.

Re: 3, yep, that also seems like a decent way of testing it.

I guess I'll just pick a system and buy it. Will probably go PCIe --> digital out --> DAC, since the PCIe should allow the quickest transfer of audio data to the sound card (as compared to USB2/3 or FireWire). Probably overkill for my current system, but will likely be what I have to do for the many-channel array. When I build that I can just move it to that rig.

**Comment by larsoner** [comment]
Actually you can probably test using onboard audio first. Some (maybe most
new?) audio cards have digital out.

**Comment by rkmaddox** [comment]
Have been doing a bit more reading about the RME cards. Seems like they get down to 48-64 samples buffer. This is worst case 1.45 ms jitter (64 samples at 44.1 kHz sampling rate). We should at some point move a photoelectrode around a monitor to see what the timing difference between different parts of the screen are. I suspect it is more than 1.45 ms.

If that's the case, 1.45 ms jitter probably doesn't matter. It would only affect AV sync, since the audio to trigger jitter should be on the order of micro or nanoseconds. The visual system seems way too slow to care about 1.45 ms jitter. The worst case I can think of is with some SSVEP / ASSR thing where there is some phase jitter of 2 ms at 20 Hz, which means 1/25th of a cycle jitter.

Probably not worth worrying about?

**Comment by rkmaddox** [comment]
And if there is an experiment where the timing between visual and auditory stimuli _really is_ that important, then we can use the soundcard (either the analog out, or the digital outs from the S/PDIF box) to drive LEDS.

**Comment by drammock** [comment]
Besides onset latency, though, smaller buffer also means more likelihood of missed frames, right?  Or does that risk basically go away with these higher-end cards?

**Comment by rkmaddox** [comment]
Yes, that's true. From what I read it seems like a normal computer can get you 48-64 samples buffer without missing frames (it depends on both card and PC load). As long as our python stuff doesn't totally suck, then it should be fine. There are people doing multitrack playing and recording with chains of effects plugins getting those latencies.

It's possible it would become a problem that we have to address when I start doing 30 output channels.

**Comment by rkmaddox** [comment]
So I just ran a test on the lab windows computer. Updated sync test to also use `sounddevice.play` to play a bipolar pulse.

Latency from TDT trigger to sound card audio was 148 +- 1.5 ms (ranged over about 3 ms). This is interesting because the block size was something like 4596 samples. The way I interpret this is that `sounddevice` is trying to make consistent latency.

The hackish code I was using opened a stream, played the sound, and closed the stream every time. It is reasonable to expect that if we opened one stream and just wrote to it that the variance could be reduced.

Promising, but still lots to figure out.

**Comment by larsoner** [comment]
> Latency from TDT trigger to sound card audio was 148 +- 1.5 ms (ranged over about 3 ms). This is interesting because the block size was something like 4596 samples. The way I interpret this is that sounddevice is trying to make consistent latency.

The other possible interpretation is that each time you do an `open` it takes roughly 148 ms to get a new stream to the sound card set up (i.e., the consistency might come from the driver), and then you immediately fill the first buffer when telling it to play...

>  It is reasonable to expect that if we opened one stream and just wrote to it that the variance could be reduced.

... and if that is indeed the case, if you open *once* and do repeated plays using that instance, you should instead see the jitter span that entire 4596 @ 44100 = ~100 ms range.

**Comment by rkmaddox** [comment]
Very good point. I ordered the soundcard yesterday, and it can use a 48
sample buffer. So we should know more soon.

**Comment by rkmaddox** [comment]
Sound card has been delivered, and work will begin in earnest after ARO in mid-Feb. Reading a bit more about `sounddevice`, I am convinced it will be the way to go. It is based on PortAudio, which is what psych toolbox uses to get their claimed very low latency audio. PortAudio also plays nice with ASIO, which gives by far the best latencies in Windows.

So, but much will happen in the next three weeks, but then our goal is to get this thing working. We lost an experiment this weekend to a malfunctioning TDT...

Also, adding @maddycapp27 to this thread, since we'll likely be doing the coding and testing together.

**Comment by rkmaddox** [comment]
Also, putting this here for posterity: http://www.portaudio.com/docs/portaudio_sync_acmc2003.pdf

**Comment by larsoner** [comment]
> Reading a bit more about sounddevice, I am convinced it will be the way to go. It is based on PortAudio, which is what psych toolbox uses to get their claimed very low latency audio. PortAudio also plays nice with ASIO, which gives by far the best latencies in Windows.

I had a recurring thought about using callbacks with tiny buffer sizes that might be problematic, it's problem (1) I listed above. Let's say that you could in theory have a buffer size of 1 sample. This means that the callback would need to execute within 23 microseconds (1./44100.), *every* 23 microseconds. This is doable in theory because it's a very small `memcpy`. But in practice, no *other* call can occupy the interpreter for more than 23 microseconds, including probably things like `ec.flip()` (which almost certainly will), otherwise the buffer fill callback will miss.

Take that up to the 48-sample buffer size, and now we're talking about 1 ms. `ec.flip` will take this long most of the time. So might be other calls, like reading a wav file or maybe even waiting for a button press (we'll have to check). Unless each function we use is smart enough to use non-interpreter-blocking calls, then we will have problems I think.

In other words, we might not be able to use Python-interpreter callbacks with short buffers. However, if `pyo` allows non-interpreter / non-callback-based playing of audio files (or data already in memory hopefully), it wouldn't suffer from this problem.

In any case, we'll get answers from testing soon it sounds like!

**Comment by rkmaddox** [comment]
Yep, lots of unanswered questions that will only become clear once we try to get it to work I think. I don't think we _have_ to use callbacks with `sounddevice`--we just have that option.

**Comment by larsoner** [comment]
> I don't think we have to use callbacks with sounddevice--we just have that option.

The `play` wrapper is pure Python that wraps the `callback`-based API:

https://github.com/spatialaudio/python-sounddevice/blob/master/sounddevice.py#L95

So it should have the same problem.

**Comment by rkmaddox** [comment]
Ah. I think I confused it with this project in progress, which is built on `sounddevice`: https://github.com/spatialaudio/python-rtmixer. They mention a non-blocking callback function. Would that solve the problem?

**Comment by rkmaddox** [comment]
(That same project also mentions fixed latency playback)

**Comment by larsoner** [comment]
Yes it sounds like they explicitly aim to fix the problem I'm talking about! Don't bother trying `sounddevice` directly, the existence of `python-rtmixer` confirms that `sounddevice` will have this problem, so I'd start with `python-rtmixer` instead.

**Comment by rkmaddox** [comment]
Roger. Do you think it's OK to use `python-rtmixer` even though it's not mature? If we get it working, we should probably have `expyfun` download the specific working commit instead of the most recent to future-proof it. A change to the code that messed with the latency could destroy a study...

**Comment by larsoner** [comment]
Hopefully we can install it on a system once and never change it (like we do with Pyglet). We can keep a record of what version or commit numbers we used at the time. That way we don't need to package it ourselves. Any time a new recording rig is set up we can try the latest one, hopefully it works, if it doesn't we either fix it upstream or `pip install` via commit number.

**Comment by drammock** [comment]
hey @rkmaddox, what ever happened with your testing of the RME card?  Next week I'll be attempting to get expyfun up and running on a new lab system that uses a [lynx e44](https://www.lynxstudio.com/products/e44/) sound card.  Pyglet will likely be good enough for some of the experiments they do, but others will need better timing, so hence I'm resurrecting this thread.

**Comment by larsoner** [comment]
I have some work in progress code for this. I can open a WIP PR tomorrow

**Comment by drammock** [comment]
implemented in #366
